libusb/io.c

```diff
@@ -311,7 +311,11 @@ if (r == 0 && actual_length == sizeof(data)) {
  * libusb_cancel_transfer() is asynchronous/non-blocking in itself. When the
  * cancellation actually completes, the transfer's callback function will
  * be invoked, and the callback function should check the transfer status to
- * determine that it was cancelled.
+ * determine that it was cancelled. On macOS and iOS it is not possible to
+ * cancel a single transfer. In this case cancelling one tranfer on an endpoint
+ * will cause all transfers on that endpoint to be cancelled. In some cases
+ * the call may cause the endpoint to stall. A call to \ref libusb_clear_halt
+ * may be needed.
```

Comment:
I don't think this description is accurate:

> In some cases the call may cause the endpoint to stall. A call to `libusb_clear_halt` may be needed.

It's not clear to me what you mean by this call causing the endpoint to stall.

From my reading of the standard, the endpoint is only stalled/halted as a result of a `STALL` packet from the device, a bus error, or a `SetFeature(ENDPOINT_HALT)` request.

I don't think `libusb_cancel_transfer` / `AbortPipe` can cause an endpoint to be stalled, unless the call involved sending a `SetFeature(ENDPOINT_HALT)` request to the device, and I see no indication that `AbortPipe` does this.

As far as I can see, a call to `libusb_clear_halt` should only be needed if the endpoint was already known to be stalled. Calling `libusb_cancel_transfer` cannot cause it to become stalled as far as I can see.

Reply:
Maybe this has been observed on some devices, even if it is against the standard? But I agree "In some cases" is too vague. It is tempting to remove this part if we cannot be more accurate. Something like "For some devices" would be better - if it is true.

Reply:
Now after reading https://github.com/libusb/libusb/issues/1110#issuecomment-1090621211 I believe the part about endpoint stalls was based on a misunderstanding and should be removed?

Reply:
Yes that's right.